Convert the Packer configuration from JSON to HCL2

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request converts the existing Packer configuration from a [JSON template](https://www.packer.io/docs/templates/legacy_json_templates) to an [HCL2 template](https://www.packer.io/docs/templates/hcl_templates). This closes #64.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

As of Packer v1.7 HCL2 is the preferred way to write Packer configurations. Switching to HCL2 templates not only modernizes our configuration but also allows the following:
- Eventually splitting up the configuration into multiple files a la Terraform configurations.
- Comments!
- The usage of `.pkrvars.hcl` files to provide values for variables (negating the need for patching scripts).
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass. I confirmed that a pre-release AMI was built and has the correct tags.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Add a tag or create a release.

mcdonnnj committed:
Convert the Packer configuration from JSON to HCL2

This was done using the command:
packer hcl2_upgrade packer.json

mcdonnnj committed:
Fix autogenerated Packer HCL2 configuration

Make the necessary tweaks to the autogenerated configuration to pass
`packer validate`.

mcdonnnj committed:
Remove unnecessary variable interpolation

mcdonnnj committed:
Change autogenerated names to be more descriptive

mcdonnnj committed:
Move away variables through environment variables

Configure the GitHub Actions workflows to pass variables through the
command line instead of setting environment variables. This also makes
it more straightforward to use a variables file if desired.

mcdonnnj committed:
Add variable descriptions

mcdonnnj committed:
Fix usage of the amazon-ami data resource

Remove unnecessary quotes and use the data resource to retrieve the AMI
name.

mcdonnnj committed:
Add variables to the Packer configuration

Add variables to control the AMI region(s) to copy a created AMI into
and the KMS key(s) to use when copying. Since these are now variables
that can be provided through a .pkrvars.hcl file we no longer need the
patch_packer_config.py script.

mcdonnnj committed:
Update the README

Update the README to reflect the changes made to the Packer
configuration and how this impacts AMI building.

mcdonnnj committed:
Fix variable used for AWS role assumption

The variable used to specify the AWS region was not adjusted when the
BUILD_REGION variable was removed.

mcdonnnj committed:
Update README with auto-generated output

Use terraform-docs on a variables.tf containing the variables defined
in packer.pkr.hcl to get a consistently formatted table of inputs.

mcdonnnj committed:
Update the variables in the Packer configuration

Change the type of the skip_create_ami variable to better reflect what
the it represents. Update the descriptions for three variables to
reference that the values are used for AMI tagging.

mcdonnnj committed:
Bump version from 0.1.2 to 1.0.0-rc.1

dav3r commented on src/packer.pkr.hcl:
```diff
@@ -0,0 +1,131 @@
+variable "is_prerelease" {
+  default     = "false"
+  description = "The pre-release status to use for the tags applied to the created AMI."
+  type        = string
```
Shouldn't this be a boolean?
```suggestion
  type        = bool
```

mcdonnnj replied:
We usually do `map(string)` for AWS tags, so I thought it should be left as a string since it is only used for tagging. If desired I can look at converting it to a `bool`, but that will require some massaging of the workflows to avoid empty input values.

dav3r replied:
If it's a big hassle, I'm fine leaving it as a string, but in that case, we should mention in the variable description that valid values are "true" and "false".

jsf9k replied:
I vote for changing it to a `bool` unless it's a really heavy lift.  As an input variable it is user-facing and therefore should be a simple as possible.  A `bool` is definitely simpler (and less error prone) than a `string` that takes on the two values `true` and `false`.

If it is a really heavy lift and we need to leave it as a `string` then please add a comment explaining why that is the case; otherwise in a month I will have no memory of this interaction and I will suggest changing it to a `bool` again.  Also, if the variable must remain a string then please make use of [validation rules](https://www.packer.io/docs/templates/hcl_templates/variables#custom-validation-rules) to ensure that `true` and `false` are really the only values that will be accepted.

mcdonnnj replied:
The main issue is in the workflows the value of `github.event.release.prerelease` is empty if it isn't a pre-release/release. We need to ensure it's a valid string version of a boolean value before it can be passed in to the command line. The quick lift is to remove all references to `github.event.release` in the `build.yml` workflow because it is both not creating an AMI and not designed to process a pre-release/release, so you end up with

```console
  packer build -timestamp-ui \
    -var is_prerelease= \
    -var release_tag= \
    -var release_url= \
    -var skip_create_ami=true \
    src/packer.pkr.hcl
```

when the workflow runs (per the [logs](https://github.com/cisagov/skeleton-packer/runs/5187675424?check_suite_focus=true#step:13:2-7)). Instead alternate values could be inserted/defaults allowed to be used.

jsf9k replied:
OK, then leave it as a string but add the validation rules to ensure that only `"true"` and `"false"` are accepted for the variable.  Also make sure that the variable description tells the user what the accepted values are, including issues of case (`"true"` vs `"True"`).

dav3r replied:
What @jsf9k said works for me.

mcdonnnj replied:
Adding validation or switching to a `bool` will both require adjusting the workflow because neither allows an empty value to be passed. I would rather switch to a `bool` and remove the references to the `github.event.release` object from the `build.yml` workflow. No image is created so setting tag values is not necessary (the defaults are fine). A bool friendly value should be passed in when the `prerelease.yml` and/or `release.yml` workflow run because they should have a `github.event.release` object.

jsf9k replied:
Why does adding validation require adjusting the workflow?  Can't you just use [this](https://www.packer.io/docs/templates/hcl_templates/variables#custom-validation-rules)?

jsf9k replied:
Never mind, it's the empty values that are the problem.

jsf9k replied:
> I would rather switch to a `bool` and remove the references to the `github.event.release` object from the `build.yml` workflow. No image is created so setting tag values is not necessary (the defaults are fine). A bool friendly value should be passed in when the `prerelease.yml` and/or `release.yml` workflow run because they should have a `github.event.release` object.

I like this solution.  Can you either do this or create an issue as a reminder to do it later?

mcdonnnj replied:
Please see [d159adc](https://github.com/cisagov/skeleton-packer/pull/98/commits/d159adcfe324de818da7a0e0f440807072a2b95b). I created a new pre-release and confirmed that an AMI was successfully created and that the tags are correctly set on the AWS console.

dav3r replied:
That looks much better! 👍

dav3r replied:
@jsf9k - Let us know if https://github.com/cisagov/skeleton-packer/pull/98/commits/d159adcfe324de818da7a0e0f440807072a2b95b works for you or if you have any concerns.  Once I hear from you, I'll resolve this conversation (or feel free to just resolve it yourself).

dav3r commented on README.md:
````diff
@@ -104,46 +104,50 @@ source_profile = build-skeleton-packer
-
 ```console
-./patch_packer_config.py --help
+AWS_PROFILE=cool-images-ec2amicreate-skeleton-packer packer build --timestamp-ui -var-file release.pkrvars.hcl src/packer.pkr.hcl
````
Did you verify that if you follow these updated instructions, you can successfully build an AMI in one region and have it copied to one or more other regions?

mcdonnnj replied:
I have used these directions to create an AMI and have it copy to an additional region.

mcdonnnj committed:
Change is_prerelease variable type

Switch from a string to a bool variable type to better reflect what the
variable represents. This also requires updating the build.yml GitHub
Actions workflow to remove variable value passing in the call to
Packer. Since this workflow does not create an AMI we do not need to
worry about the tag values since they will not be applied. The
github.event.release object also contains no useful values in that
workflow's context. The correct bool-like value for is_prerelease
should be passed in the prerelease.yml and release.yml workflows.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Bump version from 1.0.0-rc.1 to 1.0.0-rc.2

dav3r reviewed (approved):
Nice work on this! 👽

jsf9k reviewed (approved):
Excellent work!

mcdonnnj committed:
Merge branch 'develop' into improvement/use_hcl2
